### PR TITLE
Correct owner chains when mixing byname implicits and inlining

### LINKED ISSUE #5766: Mixing by-name implicits and inlining results in broken owner chains
The following,
```scala
trait Foo { def next: Foo }

inline implicit def foo(implicit loop: => Foo): Foo = new Foo { def next = loop }

def summon(implicit f: Foo) = ()
summon
```
crashes the compiler with,
```
Exception in thread "main" java.lang.AssertionError: assertion failed: failure to
construct path from method next/anonymous class Object with Test1.Foo{...}/value
<local Test1$>/object Test1/package <empty>/package <root> to `this` of
class $_lazy_implicit_$2;
anonymous class Object with Test1.Foo{...} does not have an outer accessor
```
This is the inlining equivalent of a similar interaction between by-name implicits and macros in scalac and reported and fixed here: https://github.com/scala/scala/pull/7199.

The Dotty fix is quite similar ... PR incoming.

### PR DESCRIPTION
**Pull request by milessabin**
When constructing the dictionary for a byname implicit resolution we have to ensure that the owners of any definitions which have been hoisted into the dictionary are correct. Prior to this PR this was only
done fully for implicit expansions which didn't involve inlining. We fix that here by ensuring that all definitions under a given dictionary entry are owned directly or indirectly by that dictionary entry.

Fixes #5766.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/ast/tpd.scala [hunk 1]
```diff
@@ -730,6 +730,28 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
...
+     * Set the owner of every definition in this tree which is not itself contained in this
+     * tree to be `newowner`
+     */
+    def changeNonLocalOwners(newowner: Symbol)(implicit ctx: Context): Tree = {
+      val localOwners = mutable.HashSet[Symbol]()
+      val traverser = new TreeTraverser {
+
+        def traverse(tree: Tree)(implicit ctx: Context): Unit = {
+          tree match {
+            case _: DefTree => if(tree.symbol ne NoSymbol) localOwners += tree.symbol
```
[the remark this answers is not in the document]
Comment (milessabin) [reply compiler/src/dotty/tools/dotc/ast/tpd.scala]: It seems not.

[the remark this answers is not in the document]
Comment (milessabin) [reply compiler/src/dotty/tools/dotc/ast/tpd.scala]: I've pushed a change only continuing in the non-`DefTree` case.

File: compiler/src/dotty/tools/dotc/ast/tpd.scala [hunk 2]
```diff
@@ -730,6 +730,27 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
...
+            case _ =>          traverseChildren(tree)
+          }
+        }
+      }
+      traverser.traverse(tree)
+      val nonLocalOwners = localOwners.filter(sym => !localOwners.contains(sym.owner)).toList.map(_.owner)
+      val newOwners = nonLocalOwners.map(_ => newowner)
+      new TreeTypeMap(oldOwners = nonLocalOwners, newOwners = newOwners).apply(tree)
+    }
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/ast/tpd.scala]: Can this be simplified further?
```scala
def changeNonLocalOwners(newOwner: Symbol)(implicit ctx: Context): Tree = {
  val localOwnerAcc = new TreeAccumulator[List[Symbol]] {
    def apply(ss: List[Symbol], tree: Tree)(implicit ctx: Context) = tree match {
      case tree: DefTree =>
        if (tree.symbol.exists) tree.symbol :: ss
        else ss
      case _ =>
        foldOver(ss, tree)
      }
    }
  }
  val nonLocalOwners = localOwnerAcc(Nil, tree).map(_.owner).distinct
  val newOwners = nonLocalOwners.map(_ => newOwner)
  new TreeTypeMap(oldOwners = nonLocalOwners, newOwners = newOwners).apply(tree)
}
```

Comment (odersky) [reply compiler/src/dotty/tools/dotc/ast/tpd.scala]: I think the idea of using an accumulator is good, but why the detour via lists? Let's use an accumulator over an immutable set instead.

Comment (allanrenucci) [reply compiler/src/dotty/tools/dotc/ast/tpd.scala]: Updated:
```scala
def changeNonLocalOwners(newOwner: Symbol)(implicit ctx: Context): Tree = {
  val ownerAcc = new TreeAccumulator[List[Symbol]] {
    def apply(ss: Set[Symbol], tree: Tree)(implicit ctx: Context) = tree match {
      case tree: DefTree =>
        if (tree.symbol.exists) tree.symbol + ss.owner
        else ss
      case _ =>
        foldOver(ss, tree)
      }
    }
  }
  val owners = ownerAcc(Nil, tree)
  val newOwners = List.fill(owners.size)(newOwner)
  new TreeTypeMap(oldOwners = owners, newOwners = newOwners).apply(tree)
}
```

**Pull request merged**